dartsim/src/JointFeatures_TEST.cc

```diff
@@ -163,6 +166,467 @@ TEST_F(JointFeaturesFixture, JointSetCommand)
+  EXPECT_NEAR(pos, joint->GetPosition(0), 1e-2);
+
+  joint->SetMinVelocityCommand(0, -1);
+  joint->SetMaxVelocityCommand(0, 1);
+
+  for (std::size_t i = 0; i < 100; ++i)
+  {
+    joint->SetForce(0, 1000);
+    world->Step(output, state, input);
+  }
+  EXPECT_NEAR(pos + 0.1, joint->GetPosition(0), 1e-2);
+  EXPECT_NEAR(1, joint->GetVelocity(0), 1e-6);
+
+  joint->SetPosition(0, pos);
+  EXPECT_NEAR(pos, joint->GetPosition(0), 1e-2);
+
+  joint->SetMinPositionCommand(0, -1e6);
+  joint->SetMaxPositionCommand(0, 1e6);
+
+  for (std::size_t i = 0; i < 100; ++i)
+  {
+    joint->SetForce(0, 1000);
+    world->Step(output, state, input);
+  }
+  EXPECT_LT(pos + 0.1, joint->GetPosition(0));
+  EXPECT_NEAR(1, joint->GetVelocity(0), 1e-6);
+}
+
+// TODO(anyone): position limits do not work very well with velocity control
+// bug https://github.com/dartsim/dart/issues/1583
+#if 0
+TEST_F(JointFeaturesFixture, JointSetPositionLimitsWithVelocityControl)
+{
+  sdf::Root root;
+  const sdf::Errors errors = root.Load(TEST_WORLD_DIR "test.world");
+  ASSERT_TRUE(errors.empty()) << errors.front();
+
+  const std::string modelName{"simple_joint_test"};
+  const std::string jointName{"j1"};
+
+  auto world = this->engine->ConstructWorld(*root.WorldByIndex(0));
+
+  auto model = world->GetModel(modelName);
+  auto joint = model->GetJoint(jointName);
+
+  physics::ForwardStep::Output output;
+  physics::ForwardStep::State state;
+  physics::ForwardStep::Input input;
+
+  world->Step(output, state, input);
+
+  auto pos = joint->GetPosition(0);
+
+  joint->SetMinPositionCommand(0, pos - 0.1);
+  joint->SetMaxPositionCommand(0, pos + 0.1);
+  for (std::size_t i = 0; i < 1000; ++i)
+  {
+    joint->SetVelocityCommand(0, 1);
+    world->Step(output, state, input);
+
+    if (i % 500 == 499)
+    {
+      EXPECT_NEAR(pos + 0.1, joint->GetPosition(0), 1e-2);
+      EXPECT_NEAR(0, joint->GetVelocity(0), 1e-6);
+    }
+  }
+}
+#endif
+
+TEST_F(JointFeaturesFixture, JointSetVelocityLimitsWithVelocityControl)
+{
+  sdf::Root root;
+  const sdf::Errors errors = root.Load(TEST_WORLD_DIR "test.world");
+  ASSERT_TRUE(errors.empty()) << errors.front();
+
+  auto world = this->engine->ConstructWorld(*root.WorldByIndex(0));
+  auto model = world->GetModel("simple_joint_test");
+  auto joint = model->GetJoint("j1");
+
+  physics::ForwardStep::Output output;
+  physics::ForwardStep::State state;
+  physics::ForwardStep::Input input;
+
+  joint->SetMinVelocityCommand(0, -0.1);
+  joint->SetMaxVelocityCommand(0, 0.1);
+
+  for (std::size_t i = 0; i < 100; ++i)
+  {
+    joint->SetVelocityCommand(0, 1);
+    world->Step(output, state, input);
+  }
+  EXPECT_NEAR(0.1, joint->GetVelocity(0), 1e-6);
+
+  for (std::size_t i = 0; i < 10; ++i)
+  {
+    joint->SetVelocityCommand(0, 0.1);
+    world->Step(output, state, input);
+  }
+  EXPECT_NEAR(0.1, joint->GetVelocity(0), 1e-6);
+
+  for (std::size_t i = 0; i < 10; ++i)
+  {
+    joint->SetVelocityCommand(0, -0.025);
```

Comment:
Should this be less than `-0.1` to test the lower limit properly?

Reply:
I've kept the `-0.025` test and added a `-1` test. I wanted the `-0.025` test to verify that the limit isn't "overapplied" and doesn't change proper velocity commands.